test/e2e/common/util.go

```diff
@@ -184,3 +186,60 @@ func RestartNodes(c clientset.Interface, nodes []v1.Node) error {
 	}
 	return nil
 }
+
+func RecreateNodes(c clientset.Interface, nodes []v1.Node) error {
+	// Build mapping from zone to nodes in that zone.
+	nodeNamesByZone := make(map[string][]string)
+	for i := range nodes {
+		node := &nodes[i]
+		zone := framework.TestContext.CloudConfig.Zone
+		if z, ok := node.Labels[v1.LabelZoneFailureDomain]; ok {
+			zone = z
+		}
+		nodeNamesByZone[zone] = append(nodeNamesByZone[zone], node.Name)
+	}
+
+	// Find the sole managed instance group name
+	var instanceGroup string
+	if strings.Index(framework.TestContext.CloudConfig.NodeInstanceGroup, ",") >= 0 {
+		framework.Failf("Test does not support cluster setup with more than one managed instance group: %s", framework.TestContext.CloudConfig.NodeInstanceGroup)
+	} else {
+		instanceGroup = framework.TestContext.CloudConfig.NodeInstanceGroup
+	}
+
+	// Recreate the nodes.
+	for zone, nodeNames := range nodeNamesByZone {
+		args := []string{
+			"compute",
+			fmt.Sprintf("--project=%s", framework.TestContext.CloudConfig.ProjectID),
+			"instance-groups",
+			"managed",
+			"recreate-instances",
+			instanceGroup,
+		}
+
+		args = append(args, fmt.Sprintf("--instances=%s", strings.Join(nodeNames, ",")))
+		args = append(args, fmt.Sprintf("--zone=%s", zone))
+		framework.Logf("Recreating instance group %s.", instanceGroup)
+		stdout, stderr, err := framework.RunCmd("gcloud", args...)
+		if err != nil {
+			return fmt.Errorf("error restarting nodes: %s\nstdout: %s\nstderr: %s", err, stdout, stderr)
+		}
+	}
+
+	// Wait for their boot IDs to change.
+	for i := range nodes {
+		node := &nodes[i]
+		if err := wait.Poll(pollInterval, framework.RecreateNodeReadyAgainTimeout, func() (bool, error) {
+			newNode, err := c.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+			if err != nil {
+				framework.Logf("Could not get node info: %s. Retrying in %v.", err, pollInterval)
+				return false, nil
+			}
+			return node.Status.NodeInfo.BootID != newNode.Status.NodeInfo.BootID, nil
+		}); err != nil {
+			return fmt.Errorf("error waiting for node %s boot ID to change: %s", node.Name, err)
```

Comment:
Confused about this section (made another comment about who/when/how we might want to do polling to be more flexible/clear) and the messaging from the errors.

 - In this error message with return an error at the first node that doesn't change its boot ID, but `error waiting for node X...` so it makes it sound like it was maybe the only one that failed?
 - In the test you take this error and report that it `failed to recreate at least one node`, but in reality could actually be that _all but one_ of the nodes recreated successfully but the first node you tested failed.

Reply:
I'm not sure if we want to fail fast here or to wait for all nodes to be checked and emit a message indicating all the nodes that failed. Currently, I changed it to check all nodes and return an error with all that failed.

Your second point is the original intention in this case, since it was trying to fail as soon as one node wasn't fully restarted in the timeout.

Reply:
So I think I just realized a confusion in the language; I was interpreting it as:
"failed to recreate at least one node" -> I couldn't even recreate a single node
and what it sounds like you meant was
"failed to recreate at least one node" -> there is at least one node (maybe more) that failed to recreate